### ISSUE TITLE
Decide a (versioning) policy for packages of upstream projects

### ISSUE THREAD

**Issue by joseph-wakeling-sociomantic** [sociomantic-tsunami/neptune#145]
Sometimes it may be convenient to create our own packages of 3rd-party projects which we do not control.  This might be because there are no pre-existing packages to use, or because we have some special requirements that are not met by existing packages.

There are several things that a neptune policy on this might need to cover, including:

* how to maintain the package

  - probably a good baseline is to maintain a separate "package" branch, forked from the upstream release, to contain our extra patches; we can then merge regularly from upstream as new patches or releases land (see e.g. http://honk.sigxcpu.org/projects/git-buildpackage/manual-html/gbp.intro.html for a similar practice)

* how to version the package

  - we can reasonably assume that we might release several different package versions of the same upstream release; we need a way to indicate this in the package version number

The second is the more urgent need for now, given that we are wanting to release a new package for `libmxnet` in the near future.

Package versions should obviously follow the upstream project's versioning scheme (which may not be SemVer compatible), but we would like to be able to add some extra version information reflecting the _package_ version and stability.  For example, we might want to have multiple alpha, beta and release-candidate releases of each package version.  However, alpha/beta/rc _package_ versions must be clearly distinguishable from packages of upstream alpha/beta/rc releases.

**Comment by joseph-wakeling-sociomantic** [comment]
The current proposal is that a _package version_ should be supplied in the form `supplier.N` where `supplier` is some name of whoever is doing the packaging (e.g. `sociomantic`, or `tsunami`) and `N` is the package version (1, 2, 3, ...).

For stable package releases, this could be added as version metadata, i.e. with a complete version of the form `{upstream_version}+supplier.N`: for example, `v2.3.4+sociomantic.1` for Sociomantic package 1 of upstream version 2.3.4.

For package pre-releases, the supplier and package version could be rolled into pre-release info, e.g.: `v2.3.4-sociomantic.1.alpha.1`.  (Note that for example v2.3.4-sociomantic.2.alpha.1 is a higher version number than e.g. v2.3.4-sociomantic.1.rc.5.)

This would then clearly be distinguishable from a version like `v2.3.5-alpha.1+sociomantic.1` (i.e. a stable sociomantic package of an upstream alpha).

**Comment by joseph-wakeling-sociomantic** [comment]
Examples of how tags translate into debian versions with current makd:

* pre-release package of stable upstream release: tag `v1.0.0-tsunami.1.alpha.1` => deb package version `1.0.0~tsunami.1.alpha.1-xenial`

* stable package of upstream pre-release: tag `v1.0.0-alpha.1+tsunami.1` => deb package version `1.0.0~alpha.1~tsunami.1-xenial`

* pre-release package of upstream pre-release: tag `v1.0.0-alpha.1-tsunami.2.rc.3` => deb package version `1.0.0~alpha.1.tsunami.2.rc.3-xenial`

 - no difference if e.g. we were to tag `v1.0.0-alpha.1.tsunami.2.rc.3`

* stable package of stable upstream release: tag `v1.0.0+tsunami.2` => deb package version `1.0.0~tsunami.2-xenial`.

**Comment by joseph-wakeling-sociomantic** [comment]
(This is starting to feel a bit over-engineered, but let's at least talk it through ...)

**Comment by leandro-lucarella-sociomantic** [comment]
> The second is the more urgent need for now, given that we are wanting to release a new package for libmxnet in the near future.

I wonder if the first point even belong to Neptune (it might, I'm just not sure and it might be worth to discuss separately).

**Comment by leandro-lucarella-sociomantic** [comment]
> Package versions should obviously follow the upstream project's versioning scheme (which may not be SemVer compatible), but we would like to be able to add some extra version information reflecting the package version and stability.

Here we have a conflict, because Neptune is explicitly a guideline on how to implement SemVer. So if the upstream project doesn't use SemVer, it is really not compatible with Neptune. I would suggest for now to stay in the SemVer world and assume the upstream project uses SemVer.

**Comment by leandro-lucarella-sociomantic** [comment]
BTW, if the upstream project is not SemVer compatible and your current use case is only about packaging (i.e. you never touch upstream), then I don't see any reasons not to just use Debian versions as tags directly, and not even go to the SemVer world. Debian explicitly covers the case of a release of an upstream software.

**Comment by leandro-lucarella-sociomantic** [comment]
So, to summarize I see this proposal flawed for the following reasons:

* It's incompatible with Neptune. Neptune is a guideline to implement SemVer, so upstream versions need to be SemVer compatible, or the guideline need to provide a translation between the upstream version and a SemVer version, but since all upstream are different, there is no way to define this in a general way.

* It's incomplete. It lacks the case of a project that using an upstream project as a base but only does a few minor changes. I think this distinction is important in particular for pre-releases, because if you are packaging an upstream pre-release, and you want to make a pre-release for that, things get meta and there should be a way to represent that, or a recommendation to not do pre-releases of pre-releases for example.

I suggest you to re-think the proposal based on those flaws and see what you can come up with.

**Comment by mathias-baumann-sociomantic** [comment]
Can't you use our existing `-prerelease` and `+metadata` syntax to do what you want?

`v1.2.3-our_alpha.1+upstream.v9.84.2.their_beta.4`

**Comment by joseph-wakeling-sociomantic** [comment]
> I wonder if the first point even belong to Neptune (it might, I'm just not sure and it might be worth to discuss separately).

I agree that it can be considered separately.  This is more a Tsunami problem than a neptune one.  We could e.g. move it to https://github.com/sociomantic-tsunami/mxnet/ which is the actual concrete use-case.

> Here we have a conflict, because Neptune is explicitly a guideline on how to implement SemVer. So if the upstream project doesn't use SemVer, it is really not compatible with Neptune. I would suggest for now to stay in the SemVer world and assume the upstream project uses SemVer.

We certainly don't have to worry about non-semver for the actual use case that we have.  A point of concern though is that our tools (e.g. makd's packaging, and our CI package upload) seem to rely on the presence of tags that follow neptune versioning.

So, while we can put that aside for now, it might be a good idea to give a little consideration to how we could address this case if we need to.

>  It lacks the case of a project that using an upstream project as a base but only does a few minor changes.

I'm not quite sure I follow what you mean by this.  Our actual case is where we make _no_ changes to the upstream project, but where it's reasonable to version individual package releases for the same upstream release (because we might build the package in a different way, for example).

> I think this distinction is important in particular for pre-releases, because if you are packaging an upstream pre-release, and you want to make a pre-release for that, things get meta and there should be a way to represent that, or a recommendation to not do pre-releases of pre-releases for example.

Yea, this is a reasonable point.  We could have a policy that we never package upstream alphas/betas/release candidates.  We could still use prerelease indicators like `-package.alpha.1` and so forth.

Question: of the versions:

```
v1.0.0-package.alpha.1+tsunami.1
v1.0.0-package.alpha.2+tsunami.1
v1.0.0-package.alpha.1+tsunami.2
```

... what would the sort order be in terms of install priority?  I'm taking it that the metadata would be ignored?

If so, Maybe the trivial solution here is just to say that alpha/beta/rc numbers can only ever increase (regardless of the `+sN` metadata).  In other words we'd have to do:

```
v1.0.0-package.alpha.1+tsunami.1
v1.0.0-package.alpha.2+tsunami.1
v1.0.0-package.alpha.3+tsunami.2     // alpha gets bumped even though tsunami version changed
```

**Comment by leandro-lucarella-sociomantic** [comment]
> I agree that it can be considered separately. This is more a Tsunami problem than a neptune one. We could e.g. move it to https://github.com/sociomantic-tsunami/mxnet/ which is the actual concrete use-case.

Yeah, maybe it's better, if we can generalize it afterwards we could move it to https://github.com/sociomantic-tsunami/tsunami/issues, maybe.

> We certainly don't have to worry about non-semver for the actual use case that we have. A point of concern though is that our tools (e.g. makd's packaging, and our CI package upload) seem to rely on the presence of tags that follow neptune versioning.

This is not entirely true. Makd only provides a default that works with SemVer, but you can always provide your own `make PKGVERSION=XXX` (maybe the docs are not super explicit about this, but if this can't be overridden, it is a bug).

> I'm not quite sure I follow what you mean by this. Our actual case is where we make no changes to the upstream project, but where it's reasonable to version individual package releases for the same upstream release (because we might build the package in a different way, for example).

For this case I don't see why shouldn't we just stick to Debian versioning, it was designed exactly for this case. It has an `upstream_version` and a `debian_version` which only describes changes made to the way upstream is packaged: https://www.debian.org/doc/debian-policy/#version

> We could have a policy that we never package upstream alphas/betas/release candidates.

I don't know why this limitation should be imposed, and it might go against a real need to deploy and test an upstream RC.

Anyway, again I think we are missing focus here.

I see these cases:

* Non-semver upstream: outside the scope of Neptune. Probably the most sensible scheme, either if there are changes to upstream or not, is to stick to Debian versions (adapting it as necessary), based on the particular upstream versioning system. For example for DMD, which uses a scheme like `X.YYY.Z[-beta.B]`, just replacing `-beta` with `~beta` for pre-releases would be enough (and then adding the `debian_version` by appending `-sociomantic.1` for example).

* Semver upstream: If you only make packages, then apply the same transformations [Makd applies](https://github.com/sociomantic-tsunami/makd/pull/118) and just append to that something like `.sociomantic.1` or `sociomantic.1~alpha.1` if you need to make a pre-release package (no matter if the upstream version is a pre-release or not). Actually, if you make small changes to the upstream repo, I think this scheme should work too.

So I'm starting to realize this is not at all a Neptune issue at the end 😆 

I think for repos that are only tracking upstream repos to make packages, we just need to use Debian compatible versioning in our repo instead of SemVer versions, using SemVer only make things more complicated without any real gain (well, maybe the gain would be the possibility to use neptune on those repos). But if that's desired, I think the only general policy we can have is what [Mathias suggested](https://github.com/sociomantic-tsunami/neptune/issues/145#issuecomment-359392343).

**Comment by joseph-wakeling-sociomantic** [comment]
> For this case I don't see why shouldn't we just stick to Debian versioning, it was designed exactly for this case.

So our deb package versions would be something like:

```
1.0.0-xenial0
```

(or maybe `0xenial`)

... ?  (Note that works for both pre-releases and stable releases of the package.)

> I think for repos that are only tracking upstream repos to make packages, we just need to use Debian compatible versioning in our repo instead of SemVer versions, using SemVer only make things more complicated without any real gain (well, maybe the gain would be the possibility to use neptune on those repos).

Agree.  For package-only repos it makes sense to tag with debian-compatible versions.  But experience seems to be that our CI setup doesn't promote packages if the tag does not meet expectations.

There does remain one (not very important for now) neptune-related concern: auto-generation of GitHub releases and release notifications.  I don't think that needs to be addressed now but the more upstream packages we have, the more useful that would be to have support for.

**Comment by joseph-wakeling-sociomantic** [comment]
There would still need to be some automated component of the above in order to get the `xenial` bit in (because that's not part of the version tag, it's part of the build).

FWIW typical deb-packaging will use a version like (for vim):

```
2:7.4.1689-3ubuntu1.2
```

Here we have epoch 2, upstream release 7.4.1689, package version 3.  So I think we might reasonably go for a release tag of the form `vX.y.z-N` where N is the package version for that upstream release.

Specific distro targeting generally shouldn't be part of the _package_ version, so our own default setup is currently problematic.  `git` shows how target distros are typically included:

```
1:2.16.1-0ppa1~ubuntu16.04.1
```

Package version is `0ppa1`, and a `~` is used to add the distro.  (We've talked about this before I think:-)

So for our own projects, a good deb-package would have a version number along the lines of:

```
1.2.3-0sociomantic1~xenial
```

**Comment by joseph-wakeling-sociomantic** [comment]
Anyway, to avoid interminable discussion: AFAICS a tag like

```
v1.0.0+sociomantic~beta.1
```

... will produce the deb version associated with your suggestion (I'll test and confirm).  The main issue then will be whether the resulting tag gets handled correctly by CI when deciding whether or not to upload a generated package, and (if so) whether to place it in the release or pre-release repository.

**Comment by joseph-wakeling-sociomantic** [comment]
Looks like `~` is not a valid tag component.  But using a tag:

```
v1.0.0+sociomantic.1-rc.1
```

... results in a deb package version of

```
1.0.0~sociomantic.1.rc.1-xenial
```

... which looks a bit problematic from a deb-versioning point of view.

Tell you what: I'll see if I can develop a custom package makefile that slightly tweaks the handling of version tags, and gets us something along the lines of what we want.

**Comment by leandro-lucarella-sociomantic** [comment]
> v1.0.0+sociomantic~beta.1

I think this is a bad idea, this version string is not either compatible with Debian or SemVer, we get the worse of both worlds. :(

**Comment by leandro-lucarella-sociomantic** [comment]
Wouldn't just use a Debian version for the tag and override `PKGVERSION` in makd work? I think is the simplest solution.

**Comment by joseph-wakeling-sociomantic** [comment]
Not if you want a distro-dependent component (e.g. `xenial`) in the deb package version.  But if you're happy to drop that, then yes, I think a deb-version tag would be fine.

**Comment by leandro-lucarella-sociomantic** [comment]
I wouldn't make that a blocker, but I don't see why can't you have that too. Just use `PKGVERSION=WHATEVER~$(lsb_release -cs)`.

**Comment by joseph-wakeling-sociomantic** [comment]
Minor note: if I build with `PKGVERSION` set manually in the call to `make`, I still get the distro version appended:

```
make PKGVERSION=1.0.0-0sociomantic0~xenial pkg
```

... results in an actual package version of `1.0.0-0sociomantic0~xenial-xenial`.  Looks like `makd` is really quite opinionated about versions :-P
